bpf: fix iptables masquerading for node -> remote pod traffic

jibi opened the pull request:
When Cilium runs with KPR, host-firewall or bandwidth manager, it will
try to auto-derive one or more devices to which the bpf_host program is
attached.

This program will, among other things, redirect ingress traffic destined
to a pod to the pod's lxc device using `bpf_redirect()`.

This causes the traffic to bypass the nf_conntrack table, leading to a
situation where traffic leaving the pod after the connection's been
established will be (incorrectly) masqueraded in case Iptables
masquerading is enabled, since the connection is not tracked by
netfilter.

This commit fixes this by skipping `bpf_redirect()` when we detect this
case (i.e. traffic is flowing through bpf_host attached to a physical
device and Cilium has installed Iptables rules which require conntrack).

Fixes: https://github.com/cilium/cilium/issues/14859.
Fixes: https://github.com/cilium/cilium/issues/12205.

jibi committed:
test: re-enable K8sDatapathConfig Host firewall tests

This commit re-enables the "K8sDatapathConfig Host firewall tests With
native routing" and "K8sDatapathConfig Host firewall tests With native
routing and endpoint routes" tests to run with kube-proxy

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -526,6 +526,23 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 			return DROP_INVALID;
 	}
 
+#if defined(NO_REDIRECT) && !defined(ENABLE_REDIRECT_FAST)
```
Isn't this only an issue if BPF masquerading is disabled? Does it also happen when KPR is enabled but not BPF masquerading?

I also think the proper place to fix this would be in https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L1451-L1465.

jibi replied:
> Isn't this only an issue if BPF masquerading is disabled? Does it also happen when KPR is enabled but not BPF masquerading?

The issue is only there when both BPF masquerading and KPR are disabled (i.e. not an issue with BPF masquerading disabled and KPR enabled). Does it mean the guard above is incorrect?

> I also think the proper place to fix this would be in https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L1451-L1465.

What's the advantage of skipping redirect in `bpf_lxc`? That we would do:
```
bpf_host [netdev] -> bpf_lxc -> stack
```

rather than:
```
bpf_host [netdev] -> stack -> bpf_host [cilum_host] -> bpf_lxc
```
skipping the `bpf_host` program attached to `cilium_host`?

pchaigno replied:
> The issue is only there when both BPF masquerading and KPR are disabled (i.e. not an issue with BPF masquerading disabled and KPR enabled). Does it mean the guard above is incorrect?

I think so. At least it does not encode "BPF masquerading and KPR are disabled". For that you would need to check `ENABLE_MASQUERADING` and `ENABLE_NODEPORT` I believe.

> What's the advantage of skipping redirect in `bpf_lxc`?

Not bypassing ingress policy enforcement when endpoint routes are disabled :wink: 

`ipv4_local_delivery` will tail call to the destination pod's BPF program and enforce ingress policy. If endpoint routes are disabled, that's the only place where we enforce ingress policies.

jibi replied:
> I think so. At least it does not encode "BPF masquerading and KPR are disabled". For that you would need to check `ENABLE_MASQUERADING` and `ENABLE_NODEPORT` I believe.

ack, will look into that

> Not bypassing ingress policy enforcement when endpoint routes are disabled wink
> 
> `ipv4_local_delivery` will tail call to the destination pod's BPF program and enforce ingress policy. If endpoint routes are disabled, that's the only place where we enforce ingress policies.

I'm not sure I'm following: if endpoint routes are disabled it means that endpoint traffic should go through `cilium_host` (which will not skip redirect as `from_host` will be `true` in this context), which in turn should tail call to the destination pod's BPF program and enforce ingress policy. Am I missing something?

pchaigno replied:
> if endpoint routes are disabled it means that endpoint traffic should go through `cilium_host` (which will not skip redirect as `from_host` will be `true` in this context), which in turn should tail call to the destination pod's BPF program and enforce ingress policy. Am I missing something?

Ah, right :+1: It should work then...

brb replied:
Maybe we should just make the `#if defined(ENCAP_IFINDEX) || defined(NO_REDIRECT)` check at the line 479 in your PR independent from `#ifdef ENABLE_NODEPORT`?

jibi replied:
I run a few tests to confirm when all these macros are defined:

* `--kube-proxy-replacement disabled`:
```
root@k8s1:~# cilium status | egrep 'Masquerading|KubeProxyReplacement'
KubeProxyReplacement:   Disabled
Masquerading:           IPTables [IPv4: Enabled, IPv6: Enabled]
root@k8s1:~# egrep 'ENABLE_NODEPORT |ENABLE_MASQUERADE |NO_REDIRECT ' /var/run/cilium/state/globals/node_config.h
#define NO_REDIRECT 1
```

* `--kube-proxy-replacement strict`:
```
root@k8s1:~# cilium status | egrep 'Masquerading|KubeProxyReplacement'
KubeProxyReplacement:   Strict   [enp0s8 192.168.33.11 fd01::b (Direct Routing)]
Masquerading:           IPTables [IPv4: Enabled, IPv6: Enabled]
root@k8s1:~# egrep 'ENABLE_NODEPORT |ENABLE_MASQUERADE |NO_REDIRECT ' /var/run/cilium/state/globals/node_config.h
#define ENABLE_NODEPORT 1
#define NO_REDIRECT 1
```

*` --kube-proxy-replacement strict --enable-bpf-masquerade`:
```
root@k8s1:~# cilium status | egrep 'Masquerading|KubeProxyReplacement'
KubeProxyReplacement:   Strict   [enp0s8 192.168.33.11 fd01::b (Direct Routing)]
Masquerading:           BPF   [enp0s8]   10.11.0.0/16 [IPv4: Enabled, IPv6: Enabled]
root@k8s1:~# egrep 'ENABLE_NODEPORT |ENABLE_MASQUERADE |NO_REDIRECT ' /var/run/cilium/state/globals/node_config.h
#define ENABLE_MASQUERADE 1
#define ENABLE_NODEPORT 1
#define NO_REDIRECT 1
```

to my understanding this means we should assert both that `NO_REDIRECT` is defined as well as `ENABLE_NODEPORT` is _not_ defined, otherwise we'll skip redirect even in the second and third case (which is not optimal as we are redirecting packets back to the stack when we could avoid that)

borkmann replied:
So shouldn't this patch rather be: `bool skip_redirect = !is_defined(ENABLE_NODEPORT)` instead of `bool skip_redirect = false;` ? We already have `skip_redirect` taking care of this. Why it cannot be reused?

Also, what about v6, seems missing?

jibi replied:
>So shouldn't this patch rather be: bool skip_redirect = !is_defined(ENABLE_NODEPORT) instead of bool skip_redirect = false; ? We already have skip_redirect taking care of this. Why it cannot be reused?

If we do like that, we would skip redirect just if `ENABLE_NODEPORT` is not defined (without checking also that `NO_REDIRECT` is set)

> Also, what about v6, seems missing?

(I think I pushed the changes for v6 right before you commented on the PR)

pchaigno reviewed (changes requested):
Isn't the same fix required for the IPv6 path?

borkmann commented on bpf/bpf_host.c:
```diff
@@ -249,6 +249,12 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host)
+#if !defined(ENABLE_NODEPORT) && defined(NO_REDIRECT) && !defined(ENABLE_REDIRECT_FAST)
+	/* See comment in handle_ipv4() */
+	if (!from_host)
+		return CTX_ACT_OK;
```
What about the `skip_redirect` suggestion from https://github.com/cilium/cilium/pull/16136#discussion_r633339972 ?

brb replied:
@jibi Does this comment need to be addressed?

jibi replied:
@brb to my understanding [the same conclusion for the v4 logic](https://github.com/cilium/cilium/pull/16136#discussion_r633477903) would apply here as well (i.e. following that suggestion we would skip redirect only if `ENABLE_NODEPORT` is not defined, ignoring the `NO_REDIRECT` macro, which is not what we want)

brb commented on bpf/bpf_host.c:
```diff
@@ -526,6 +532,23 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 			return DROP_INVALID;
 	}
 
+#if !defined(ENABLE_NODEPORT) && defined(NO_REDIRECT) && !defined(ENABLE_REDIRECT_FAST)
```
OK, I think we should ignore whether `ENABLE_NODEPORT` is set or not. In the end we are interested only whether `bpf_host` is attached to a native device (i.e. `!from_host`). The following patch (untested; IPv4-only) should do the job:
```
diff --git a/bpf/bpf_host.c b/bpf/bpf_host.c
index 2feaa7053..88a2b9e95 100644
--- a/bpf/bpf_host.c
+++ b/bpf/bpf_host.c
@@ -478,6 +478,12 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
                        if (ret < 0)
                                return ret;
                }
+               /* Verifier workaround: modified ctx access. */
+               if (!revalidate_data(ctx, &data, &data_end, &ip4))
+                       return DROP_INVALID;
+       }
+#endif /* ENABLE_NODEPORT */
+
 #if defined(NO_REDIRECT) && !defined(ENABLE_REDIRECT_FAST)
                /* Without bpf_redirect_neigh() helper, we cannot redirect a
                 * packet to a local endpoint in the direct routing mode, as
@@ -488,13 +494,9 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
                 * we bypass request and reply path in the host namespace and
                 * do not run into this issue.
                 */
-               skip_redirect = true;
+               if (!from_host)
+                       skip_redirect = true;
 #endif /* NO_REDIRECT && !ENABLE_REDIRECT_FAST */
-               /* Verifier workaround: modified ctx access. */
-               if (!revalidate_data(ctx, &data, &data_end, &ip4))
-                       return DROP_INVALID;
-       }
-#endif /* ENABLE_NODEPORT */

 #ifdef ENABLE_HOST_FIREWALL
        if (from_host) {
```

brb reviewed (changes requested):
I had a second thought on this.

jibi committed:
bpf: fix iptables masquerading for node -> remote pod traffic

When Cilium runs with KPR, host-firewall or bandwidth manager, it will
try to auto-derive one or more devices to which the bpf_host program is
attached.

This program will, among other things, redirect ingress traffic destined
to a pod to the pod's lxc device using `bpf_redirect()`.

This causes the traffic to bypass the nf_conntrack table, leading to a
situation where traffic leaving the pod after the connection's been
established will be (incorrectly) masqueraded in case Iptables
masquerading is enabled, since the connection is not tracked by
netfilter.

This commit fixes this by skipping `bpf_redirect()` when we detect this
case (i.e. traffic is flowing through bpf_host attached to a physical
device and Cilium has installed Iptables rules which require conntrack).

Fixes: #14859
Suggested-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

brb reviewed (approved):
LGTM, thanks!

Would be good to get a review from @borkmann and @pchaigno too.

borkmann reviewed (approved):
LGTM, and also addresses https://github.com/cilium/cilium/pull/16136#discussion_r633434944 . Thanks!

pchaigno reviewed (approved):
LGTM! Thanks @jibi! One step closer to making the hostfw stable with kube-proxy :tada:
